### ISSUE TITLE
Hidden Abilities and Item in battle effects

### ISSUE THREAD

**Issue by EternalCode** [EternalCode/battle_engine#5]
Should this battle engine also encompass these factors? If we don't include Hidden Ability functionality inside the battle engine, it may be required for a user who wants them in their game to rewrite portions of the engine. It also speaks to completeness of the battle engine as well.

Similarly, Items have their own battle callbacks which need to be executed. Is it something this engine should handle internally, or should that be a different project. The line is quite blurry, and I think we should support these if we are aiming for a complete engine as the end product.

Potential use of branches for this may also be a solution.

**Comment by SBird1337** [comment]
I think striving for completeness is a good aproach here, the only problem I see with hidden abilities is that they would have to be supported out of combat as well. Items on the other hand would not even necessarily have to be implemented (just their effects, which the user can set on his items) - but we might offer the possibility to install a icon package with the engine (just some defines would suffice to include/exclude it)

**Comment by Touched** [comment]
I'd quite like to see these be optional features of the battle engine, hidden behind defines. Only issue I see with this is having to define them thrice (once in ARMIPS, another time for GNU AS and a third for GCC) in order to enable checking for enabled features anywhere in the code. 

If you really want to make this super complete, you can break down each feature into a separate define and group them by generation:

```c
#ifdef GENERATION_6
#define GENERATION_5
#define FEATURE_MEGA_EVOLUTION
#endif

#ifdef GENERATION_5
#define GENERATION_4
#define FEATURE_HIDDEN_ABILTIES
#endif
```

As @SBird1337 says, I don't think you need to even implement the items, just their effects (maybe hidden behind `GENERATION_X` defines). Maybe that could be in a second repository which would allow people to select exactly what items to implement and have this just contain their in-battle effects.

**Comment by EternalCode** [comment]
@Touched I'm leaning towards this solution as well. However, I think for now it should remain feature name oriented. It can be the case that a user wants Generation 6 features, but without mega evolution for example. I will strongly consider adding generation based filters for things like Moves in the near future.

So the current decision I'm making is to use Conditional Directives to toggle Item callbacks and Hidden Abilities. It will be simple to start, and later well add new items to be toggled. Thank you all for your inputs on the matter!

**Comment by Touched** [comment]
@EternalCode I was merely suggesting the `GENERATION_X` defines so that you could have very fine-grained features and group them for a better interface. If you wanted to allow generation 6 with no mega evolution you could still do that with this approach:

```c
#ifdef GENERATION_6
#define FEATURE_MEGA_EVOLUTION
#define FEATURE_GEN6_MOVES
// ...
#endif

// ...

// Provide ability to remove features
#ifdef NO_MEGA_EVOLUTION
#undef FEATURE_MEGA_EVOLUTION
#endif
```

Then compile with `-DGENERATION_6 -DNO_MEGA_EVOLUTION`.

**Comment by EternalCode** [comment]
Absolutely. I understand the utility in grouping features. This is something to consider once we decide the add more than just Items & Hidden abilities into the toggle list.
